---
title: 'Vorgehensweise: Empfangen von Ausnahmebenachrichtigungen (erste Chance)'
description: Erhalten Sie Ausnahmebenachrichtigungen (erste Chance) in .NET über das FirstChanceException-Ereignis der AppDomain-Klasse, bevor die CLR nach Ausnahmehandlern sucht.
ms.date: 03/30/2017
dev_langs:
- csharp
- vb
helpviewer_keywords:
- first-chance exception notifications
- exceptions, first chance notifications
ms.assetid: 66f002b8-a97d-4a6e-a503-2cec01689113
ms.openlocfilehash: 0b3150a52a68e078d1052a9894bb652ad35027d0
ms.sourcegitcommit: bc293b14af795e0e999e3304dd40c0222cf2ffe4
ms.translationtype: HT
ms.contentlocale: de-DE
ms.lasthandoff: 11/26/2020
ms.locfileid: "96242564"
---
# <a name="how-to-receive-first-chance-exception-notifications"></a><span data-ttu-id="1c56d-103">Vorgehensweise: Empfangen von Ausnahmebenachrichtigungen (erste Chance)</span><span class="sxs-lookup"><span data-stu-id="1c56d-103">How to: Receive First-Chance Exception Notifications</span></span>

<span data-ttu-id="1c56d-104">Mit dem <xref:System.AppDomain.FirstChanceException>-Ereignis der <xref:System.AppDomain>-Klasse erhalten Sie eine Benachrichtigung bezüglich einer ausgelösten Ausnahme, bevor die CLR mit der Suche nach Ausnahmehandlern beginnt.</span><span class="sxs-lookup"><span data-stu-id="1c56d-104">The <xref:System.AppDomain.FirstChanceException> event of the <xref:System.AppDomain> class lets you receive a notification that an exception has been thrown, before the common language runtime has begun searching for exception handlers.</span></span>

 <span data-ttu-id="1c56d-105">Das Ereignis wird auf Ebene der Anwendungsdomäne ausgelöst.</span><span class="sxs-lookup"><span data-stu-id="1c56d-105">The event is raised at the application domain level.</span></span> <span data-ttu-id="1c56d-106">Ein Ausführungsthread kann verschiedene Anwendungsdomänen durchlaufen. Es ist also möglich, dass ein Ausnahmefehler aus einer Anwendungsdomäne in einer anderen behandelt wird.</span><span class="sxs-lookup"><span data-stu-id="1c56d-106">A thread of execution can pass through multiple application domains, so an exception that is unhandled in one application domain could be handled in another application domain.</span></span> <span data-ttu-id="1c56d-107">Die Benachrichtigung findet in jeder Anwendungsdomäne statt, die einen Handler für das Ereignis hinzugefügt hat, bis eine Anwendungsdomäne die Ausnahme behandelt.</span><span class="sxs-lookup"><span data-stu-id="1c56d-107">The notification occurs in each application domain that has added a handler for the event, until an application domain handles the exception.</span></span>

 <span data-ttu-id="1c56d-108">Die Verfahren und Beispiele in diesem Artikel zeigen, wie Sie Benachrichtigungen über Ausnahmefehler der ersten Chance in einem einfachen Programm mit einer Anwendungsdomäne und in einer von Ihnen selbst erstellten Anwendungsdomäne erhalten.</span><span class="sxs-lookup"><span data-stu-id="1c56d-108">The procedures and examples in this article show how to receive first-chance exception notifications in a simple program that has one application domain, and in an application domain that you create.</span></span>

 <span data-ttu-id="1c56d-109">Ein umfangreicheres Beispiel, das mehrere Anwendungsdomänen umfasst, finden Sie im Beispiel zum <xref:System.AppDomain.FirstChanceException>-Ereignis.</span><span class="sxs-lookup"><span data-stu-id="1c56d-109">For a more complex example that spans several application domains, see the example for the <xref:System.AppDomain.FirstChanceException> event.</span></span>

## <a name="receiving-first-chance-exception-notifications-in-the-default-application-domain"></a><span data-ttu-id="1c56d-110">Empfangen von Benachrichtigungen über Ausnahmefehler der ersten Chance in Standardanwendungsdomänen</span><span class="sxs-lookup"><span data-stu-id="1c56d-110">Receiving First-Chance Exception Notifications in the Default Application Domain</span></span>

 <span data-ttu-id="1c56d-111">In der folgenden Prozedur, dem Einstiegspunkt der Anwendung (`Main()`-Methode), wird in der Standardanwendungsdomäne ausgeführt.</span><span class="sxs-lookup"><span data-stu-id="1c56d-111">In the following procedure, the entry point for the application, the `Main()` method, runs in the default application domain.</span></span>

#### <a name="to-demonstrate-first-chance-exception-notifications-in-the-default-application-domain"></a><span data-ttu-id="1c56d-112">So demonstrieren Sie Benachrichtigungen über Ausnahmefehler der ersten Chance in der Standardanwendungsdomäne</span><span class="sxs-lookup"><span data-stu-id="1c56d-112">To demonstrate first-chance exception notifications in the default application domain</span></span>

1. <span data-ttu-id="1c56d-113">Definieren Sie mithilfe einer Lambda-Funktion einen Ereignishandler für das <xref:System.AppDomain.FirstChanceException>-Ereignis, und fügen Sie ihn an das Ereignis an.</span><span class="sxs-lookup"><span data-stu-id="1c56d-113">Define an event handler for the <xref:System.AppDomain.FirstChanceException> event, using a lambda function, and attach it to the event.</span></span> <span data-ttu-id="1c56d-114">In diesem Beispiel gibt der Ereignishandler den Namen der Anwendungsdomäne, in der das Ereignis behandelt wurde, und die <xref:System.Exception.Message%2A>-Eigenschaft der Ausnahme aus.</span><span class="sxs-lookup"><span data-stu-id="1c56d-114">In this example, the event handler prints the name of the application domain where the event was handled and the exception's <xref:System.Exception.Message%2A> property.</span></span>

     [!code-csharp[System.AppDomain.FirstChanceException_howto_simple#2](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto_simple/cs/example.cs#2)]
     [!code-vb[System.AppDomain.FirstChanceException_howto_simple#2](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto_simple/vb/example.vb#2)]

2. <span data-ttu-id="1c56d-115">Lösen Sie eine Ausnahme aus, und fangen Sie sie ab.</span><span class="sxs-lookup"><span data-stu-id="1c56d-115">Throw an exception and catch it.</span></span> <span data-ttu-id="1c56d-116">Bevor die Runtime den Ausnahmehandler sucht, wird das <xref:System.AppDomain.FirstChanceException>-Ereignis ausgelöst und eine Meldung angezeigt.</span><span class="sxs-lookup"><span data-stu-id="1c56d-116">Before the runtime locates the exception handler, the <xref:System.AppDomain.FirstChanceException> event is raised and displays a message.</span></span> <span data-ttu-id="1c56d-117">Auf diese Nachricht folgt die Nachricht, die vom Ausnahmehandler angezeigt wird.</span><span class="sxs-lookup"><span data-stu-id="1c56d-117">This message is followed by the message that is displayed by the exception handler.</span></span>

     [!code-csharp[System.AppDomain.FirstChanceException_howto_simple#3](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto_simple/cs/example.cs#3)]
     [!code-vb[System.AppDomain.FirstChanceException_howto_simple#3](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto_simple/vb/example.vb#3)]

3. <span data-ttu-id="1c56d-118">Lösen Sie eine Ausnahme aus, aber fangen Sie sie nicht ab.</span><span class="sxs-lookup"><span data-stu-id="1c56d-118">Throw an exception, but do not catch it.</span></span> <span data-ttu-id="1c56d-119">Bevor die Runtime nach einem Ausnahmehandler sucht, wird das <xref:System.AppDomain.FirstChanceException>-Ereignis ausgelöst und eine Meldung angezeigt.</span><span class="sxs-lookup"><span data-stu-id="1c56d-119">Before the runtime looks for an exception handler, the <xref:System.AppDomain.FirstChanceException> event is raised and displays a message.</span></span> <span data-ttu-id="1c56d-120">Da es keinen Ausnahmehandler gibt, wird die Anwendung beendet.</span><span class="sxs-lookup"><span data-stu-id="1c56d-120">There is no exception handler, so the application terminates.</span></span>

     [!code-csharp[System.AppDomain.FirstChanceException_howto_simple#4](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto_simple/cs/example.cs#4)]
     [!code-vb[System.AppDomain.FirstChanceException_howto_simple#4](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto_simple/vb/example.vb#4)]

     <span data-ttu-id="1c56d-121">Der Code aus den ersten drei Schritten dieser Prozedur stellt eine vollständige Konsolenanwendung dar.</span><span class="sxs-lookup"><span data-stu-id="1c56d-121">The code that is shown in the first three steps of this procedure forms a complete console application.</span></span> <span data-ttu-id="1c56d-122">Die Ausgabe der Anwendung variiert je nach dem Namen der ausführbaren Datei, denn der Name der Standardanwendungsdomäne besteht aus dem Namen und der Erweiterung der ausführbaren Datei.</span><span class="sxs-lookup"><span data-stu-id="1c56d-122">The output from the application varies, depending on the name of the .exe file, because the name of the default application domain consists of the name and extension of the .exe file.</span></span> <span data-ttu-id="1c56d-123">Ein Beispiel finden Sie in der folgenden Ausgabe.</span><span class="sxs-lookup"><span data-stu-id="1c56d-123">See the following for sample output.</span></span>

     [!code-csharp[System.AppDomain.FirstChanceException_howto_simple#5](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto_simple/cs/example.cs#5)]
     [!code-vb[System.AppDomain.FirstChanceException_howto_simple#5](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto_simple/vb/example.vb#5)]

## <a name="receiving-first-chance-exception-notifications-in-another-application-domain"></a><span data-ttu-id="1c56d-124">Empfangen von Benachrichtigungen über Ausnahmefehler der ersten Chance in einer anderen Anwendungsdomäne</span><span class="sxs-lookup"><span data-stu-id="1c56d-124">Receiving First-Chance Exception Notifications in Another Application Domain</span></span>

 <span data-ttu-id="1c56d-125">Wenn Ihr Programm mehr als eine Anwendungsdomäne enthält, können Sie auswählen, welche Anwendungsdomänen Benachrichtigungen erhalten.</span><span class="sxs-lookup"><span data-stu-id="1c56d-125">If your program contains more than one application domain, you can choose which application domains receive notifications.</span></span>

#### <a name="to-receive-first-chance-exception-notifications-in-an-application-domain-that-you-create"></a><span data-ttu-id="1c56d-126">So empfangen Sie Benachrichtigungen über Ausnahmefehler der ersten Chance in einer selbst erstellten Anwendungsdomäne</span><span class="sxs-lookup"><span data-stu-id="1c56d-126">To receive first-chance exception notifications in an application domain that you create</span></span>

1. <span data-ttu-id="1c56d-127">Definieren Sie einen Ereignishandler für das <xref:System.AppDomain.FirstChanceException>-Ereignis.</span><span class="sxs-lookup"><span data-stu-id="1c56d-127">Define an event handler for the <xref:System.AppDomain.FirstChanceException> event.</span></span> <span data-ttu-id="1c56d-128">In diesem Beispiel wird eine `static`-Methode verwendet (`Shared`-Methode in Visual Basic), die den Namen der Anwendungsdomäne, in der das Ereignis behandelt wurde, und die <xref:System.Exception.Message%2A>-Eigenschaft der Ausnahme ausgibt.</span><span class="sxs-lookup"><span data-stu-id="1c56d-128">This example uses a `static` method (`Shared` method in Visual Basic) that prints the name of the application domain where the event was handled and the exception's <xref:System.Exception.Message%2A> property.</span></span>

     [!code-csharp[System.AppDomain.FirstChanceException_howto#3](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/cs/example.cs#3)]
     [!code-vb[System.AppDomain.FirstChanceException_howto#3](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/vb/example.vb#3)]

2. <span data-ttu-id="1c56d-129">Erstellen Sie eine Anwendungsdomäne, und fügen Sie dem <xref:System.AppDomain.FirstChanceException>-Ereignis dieser Anwendungsdomäne den Ereignishandler hinzu.</span><span class="sxs-lookup"><span data-stu-id="1c56d-129">Create an application domain and add the event handler to the <xref:System.AppDomain.FirstChanceException> event for that application domain.</span></span> <span data-ttu-id="1c56d-130">In diesem Beispiel heißt die Anwendungsdomäne `AD1`.</span><span class="sxs-lookup"><span data-stu-id="1c56d-130">In this example, the application domain is named `AD1`.</span></span>

     [!code-csharp[System.AppDomain.FirstChanceException_howto#2](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/cs/example.cs#2)]
     [!code-vb[System.AppDomain.FirstChanceException_howto#2](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/vb/example.vb#2)]

     <span data-ttu-id="1c56d-131">Sie können dieses Ereignis in der Standardanwendungsdomäne auf dieselbe Weise behandeln.</span><span class="sxs-lookup"><span data-stu-id="1c56d-131">You can handle this event in the default application domain in the same way.</span></span> <span data-ttu-id="1c56d-132">Verwenden Sie die statische (`static`) <xref:System.AppDomain.CurrentDomain%2A?displayProperty=nameWithType>-Eigenschaft (`Shared` in Visual Basic) in `Main()`, um einen Verweis auf die Standardanwendungsdomäne abzurufen.</span><span class="sxs-lookup"><span data-stu-id="1c56d-132">Use the `static` (`Shared` in Visual Basic) <xref:System.AppDomain.CurrentDomain%2A?displayProperty=nameWithType> property in `Main()` to get a reference to the default application domain.</span></span>

#### <a name="to-demonstrate-first-chance-exception-notifications-in-the-application-domain"></a><span data-ttu-id="1c56d-133">So demonstrieren Sie Benachrichtigungen über Ausnahmefehler der ersten Chance in der Anwendungsdomäne</span><span class="sxs-lookup"><span data-stu-id="1c56d-133">To demonstrate first-chance exception notifications in the application domain</span></span>

1. <span data-ttu-id="1c56d-134">Erstellen Sie ein `Worker`-Objekt in der Anwendungsdomäne, die Sie in der vorherigen Prozedur erstellt haben.</span><span class="sxs-lookup"><span data-stu-id="1c56d-134">Create a `Worker` object in the application domain that you created in the previous procedure.</span></span> <span data-ttu-id="1c56d-135">Die `Worker`-Klasse muss öffentlich und von <xref:System.MarshalByRefObject> abgeleitet sein, wie im vollständigen Beispiel am Ende dieses Artikels dargestellt.</span><span class="sxs-lookup"><span data-stu-id="1c56d-135">The `Worker` class must be public, and must derive from <xref:System.MarshalByRefObject>, as shown in the complete example at the end of this article.</span></span>

     [!code-csharp[System.AppDomain.FirstChanceException_howto#4](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/cs/example.cs#4)]
     [!code-vb[System.AppDomain.FirstChanceException_howto#4](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/vb/example.vb#4)]

2. <span data-ttu-id="1c56d-136">Rufen Sie eine Methode des `Worker`-Objekts auf, die eine Ausnahme auslöst.</span><span class="sxs-lookup"><span data-stu-id="1c56d-136">Call a method of the `Worker` object that throws an exception.</span></span> <span data-ttu-id="1c56d-137">In diesem Beispiel wird die `Thrower`-Methode zweimal aufgerufen.</span><span class="sxs-lookup"><span data-stu-id="1c56d-137">In this example, the `Thrower` method is called twice.</span></span> <span data-ttu-id="1c56d-138">Beim ersten Aufruf lautet das Methodenargument `true`, weswegen die Methode ihre eigene Ausnahme abfängt.</span><span class="sxs-lookup"><span data-stu-id="1c56d-138">The first time, the method argument is `true`, which causes the method to catch its own exception.</span></span> <span data-ttu-id="1c56d-139">Beim zweiten Aufruf lautet das Argument `false`, und die `Main()`-Methode fängt die Ausnahme in der Standardanwendungsdomäne ab.</span><span class="sxs-lookup"><span data-stu-id="1c56d-139">The second time, the argument is `false`, and the `Main()` method catches the exception in the default application domain.</span></span>

     [!code-csharp[System.AppDomain.FirstChanceException_howto#6](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/cs/example.cs#6)]
     [!code-vb[System.AppDomain.FirstChanceException_howto#6](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/vb/example.vb#6)]

3. <span data-ttu-id="1c56d-140">Platzieren Sie Code in der `Thrower`-Methode, um zu steuern, ob die Methode ihre eigene Ausnahme behandelt.</span><span class="sxs-lookup"><span data-stu-id="1c56d-140">Place code in the `Thrower` method to control whether the method handles its own exception.</span></span>

     [!code-csharp[System.AppDomain.FirstChanceException_howto#5](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/cs/example.cs#5)]
     [!code-vb[System.AppDomain.FirstChanceException_howto#5](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/vb/example.vb#5)]

## <a name="example"></a><span data-ttu-id="1c56d-141">Beispiel</span><span class="sxs-lookup"><span data-stu-id="1c56d-141">Example</span></span>

 <span data-ttu-id="1c56d-142">Im folgenden Beispiel wird eine Anwendungsdomäne namens `AD1` erstellt und der Anwendungsdomäne des <xref:System.AppDomain.FirstChanceException>-Ereignisses ein Ereignishandler hinzugefügt.</span><span class="sxs-lookup"><span data-stu-id="1c56d-142">The following example creates an application domain named `AD1` and adds an event handler to the application domain's <xref:System.AppDomain.FirstChanceException> event.</span></span> <span data-ttu-id="1c56d-143">Außerdem wird eine Instanz der `Worker`-Klasse in der Anwendungsdomäne erstellt und eine Methode namens `Thrower` aufgerufen, die <xref:System.ArgumentException> auslöst.</span><span class="sxs-lookup"><span data-stu-id="1c56d-143">The example creates an instance of the `Worker` class in the application domain, and calls a method named `Thrower` that throws an <xref:System.ArgumentException>.</span></span> <span data-ttu-id="1c56d-144">Je nach dem Wert des Arguments fängt die Methode die Ausnahme entweder ab, oder es tritt ein Fehler bei der Behandlung auf.</span><span class="sxs-lookup"><span data-stu-id="1c56d-144">Depending on the value of its argument, the method either catches the exception or fails to handle it.</span></span>

 <span data-ttu-id="1c56d-145">Jedes Mal, wenn die `Thrower`-Methode eine Ausnahme in `AD1` auslöst, wird das <xref:System.AppDomain.FirstChanceException>-Ereignis in `AD1` ausgelöst, und der Ereignishandler zeigt eine Meldung an.</span><span class="sxs-lookup"><span data-stu-id="1c56d-145">Each time the `Thrower` method throws an exception in `AD1`, the <xref:System.AppDomain.FirstChanceException> event is raised in `AD1`, and the event handler displays a message.</span></span> <span data-ttu-id="1c56d-146">Die Runtime sucht nach einem Ausnahmehandler.</span><span class="sxs-lookup"><span data-stu-id="1c56d-146">The runtime then looks for an exception handler.</span></span> <span data-ttu-id="1c56d-147">Im ersten Fall befindet sich der Ausnahmehandler in `AD1`.</span><span class="sxs-lookup"><span data-stu-id="1c56d-147">In the first case, the exception handler is found in `AD1`.</span></span> <span data-ttu-id="1c56d-148">Im zweiten Fall wird die Ausnahme in `AD1` nicht behandelt und stattdessen in der Standardanwendungsdomäne abgefangen.</span><span class="sxs-lookup"><span data-stu-id="1c56d-148">In the second case, the exception is unhandled in `AD1`, and instead is caught in the default application domain.</span></span>

> [!NOTE]
> <span data-ttu-id="1c56d-149">Der Name der Standardanwendungsdomäne ist identisch mit dem Namen der ausführbaren Datei.</span><span class="sxs-lookup"><span data-stu-id="1c56d-149">The name of the default application domain is the same as the name of the executable.</span></span>

 <span data-ttu-id="1c56d-150">Wenn Sie der Standardanwendungsdomäne einen Handler für das <xref:System.AppDomain.FirstChanceException>-Ereignis hinzufügen, wird das Ereignis ausgelöst und behandelt, bevor die Standardanwendungsdomäne die Ausnahme behandelt.</span><span class="sxs-lookup"><span data-stu-id="1c56d-150">If you add a handler for the <xref:System.AppDomain.FirstChanceException> event to the default application domain, the event is raised and handled before the default application domain handles the exception.</span></span> <span data-ttu-id="1c56d-151">Um dies zu veranschaulichen, fügen Sie den C#-Code `AppDomain.CurrentDomain.FirstChanceException += FirstChanceException;` (in Visual Basic: `AddHandler AppDomain.CurrentDomain.FirstChanceException, FirstChanceException`) am Anfang von `Main()` ein.</span><span class="sxs-lookup"><span data-stu-id="1c56d-151">To see this, add the C# code `AppDomain.CurrentDomain.FirstChanceException += FirstChanceException;` (in Visual Basic, `AddHandler AppDomain.CurrentDomain.FirstChanceException, FirstChanceException`) at the beginning of `Main()`.</span></span>

 [!code-csharp[System.AppDomain.FirstChanceException_howto#1](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/cs/example.cs#1)]
 [!code-vb[System.AppDomain.FirstChanceException_howto#1](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/vb/example.vb#1)]

## <a name="see-also"></a><span data-ttu-id="1c56d-152">Siehe auch</span><span class="sxs-lookup"><span data-stu-id="1c56d-152">See also</span></span>

- <xref:System.AppDomain.FirstChanceException>
